### ISSUE TITLE
Armor and accessories

### ISSUE THREAD

**Issue by IntegratedQuantum** [PixelGuys/Cubyz#604]
They need slots in the inventory and a way to craft/obtain them.

**Comment by careeoki** [comment]
How will armor be displayed on the player? When users can create custom player models, how do we ensure armor is displayed properly?

**Comment by IntegratedQuantum** [comment]
The model would require some metadata of where the head is, the eyes, and so on. This information can be used to determine where the armor is.
The armor texture has to be generated procedurally, based on the armor material, similar to how tools work right now.

**Comment by codemob-dev** [comment]
As I mentioned in #89, what if there was no distinction between armor and accessories, so we could have specific slots like ring, helmet, goggles, chest, arms, legs, etc.

**Comment by IntegratedQuantum** [comment]
I agree, it may make sense to treat them in the same way internally. The only real difference is how they map onto the body.

**Comment by ikabod-kee** [comment]
So, instead of armor being some plates on a character, how about we just have them be orbs or some sort of protective chakra? It would prevent armor from obscuring player skins and would solve all of our issues with visual armor display.

**Comment by ikabod-kee** [comment]
We could also have shields

**Comment by Argmaster** [comment]
Since ikabod likes that idea, I feel obligated to oppose it.
That said, trinket system could be neat, like in MOBA games.
The only issue is that it either hides how strong your opponent is, or pollutes the world around you.
Also I really like armor and stuff and I am not an artist unfortunately, am I stuck with default snale indefinitely?

**Comment by careeoki** [comment]
Procedural armor would probably let you heavily customize your look. It would be fun to see what wacky armor people make.

**Comment by ikabod-kee** [comment]
<img width="224" height="338" alt="Image" src="https://github.com/user-attachments/assets/7af744af-6bd7-44a9-903d-737ab6987af1" />
I was thinking it'd be some sort of particle around your character.

**Comment by ikabod-kee** [comment]
<img width="215" height="330" alt="Image" src="https://github.com/user-attachments/assets/97a08f9a-b9d8-4eae-829a-b0101cd3eb80" />
Someone also suggested greek letters

**Comment by IntegratedQuantum** [comment]
> So, instead of armor being some plates on a character, how about we just have them be orbs or some sort of protective chakra?

I don't like the amount of magic it adds. Plus I think armor being on the body could add some interesting strategic value to the game, 
Can you discover based on the enemy's armor what weapon they are weak against?
Can you make an armor that tricks your enemy into thinking it has a certain weakness?

**Comment by BoySanic** [comment]
Which accessory slots should exist?

Should they be specific jewlery slots? Generic accessory slots similar to Terraria? Mix of both?

**Comment by IntegratedQuantum** [comment]
I was thinking semi-specific slots, so you wouldn't use 10 of the most optimal ring in all slots.
(e.g. 2 ring slots, boots slot, necklace slot)
Then in each category you'd still have different accessory blueprints/shapes you could use. E.g. there could be a necklace type that only allows using nature materials for healing.

**Comment by Wunka** [comment]
I think I will start working on this:
My steps would be for now:

- [ ] create new Item Type so Armor and Accessories can be procedual crafted and have their own modifiers / stats (maybe just use the Tool Item type and make it to something more generic)
- [ ] create new matrixes for differnet accessories (starting maybe with a ring to test)
- [ ] for now let accessories be crafted in the workbench, but later should probably be moved to other workstations
- [ ] create new modifier types ( I will let others do that, I have no idea how to balance that, I just want to do the general structure)
- [ ] create an accessorie inventory with the slots in the inventory window (maybe new window that is opened like the crafiting one?)

**Comment by Argmaster** [comment]
> create new Item Type so Armor and Accessories can be procedual crafted and have their own modifiers / stats (maybe just use the Tool Item type and make it to something more generic)

I would say we should rename it to Proceduraltem and create another callback system with on right-click onLeftClick and onTick callbacks (I am not sure if I am serious here tbh, maybe partially)

**Comment by Wunka** [comment]
> I would say we should rename it to Proceduraltem and create another callback system with on right-click onLeftClick and onTick callbacks (I am not sure if I am serious here tbh, maybe partially)

I think I will atleast use the Tool struct.
So my first PR will probably be to seperate 
https://github.com/PixelGuys/Cubyz/blob/3a7b55815cf4f57670c325f54ff3751dea7a9d2b/src/items.zig#L637-L655
into a sub union of maybe the tool type or something else (can you say in zig that mutliple things have the same struct inside a union? I think not...)

**Comment by Wunka** [comment]
Maybe I coud use the same system that is used in the modfieres with the `Data`? Then modders would be easy able to add new Proceduraltems

**Comment by Argmaster** [comment]
Hm right but we have those arbitrary `swingSpeed`, `damage` etc. fields there that make no sense for accessories...

**Comment by Wunka** [comment]
> Hm right but we have those arbitrary `swingSpeed`, `maxDurability` etc. fields there that make no sense for accessories...

Thats why I want to seperate it into something like the `Data` packed struct from Modifers (It think that could be the best system with added callbacks like #2778) 
The `Data` of accessories would then have padding but that would be ok

**Comment by Argmaster** [comment]
We could rename them to `cooldown` and `power` to make it generic in code, but we would still have to use proper wording depending on what the item actually does so to not confuse players

**Comment by Wunka** [comment]
I think I will have to read into it more, if my idea is even possible. But I would also be ok with the generic naming idea or maybe a `union(type) { pickaxe: {damage: ...`

**Comment by Argmaster** [comment]
We were looking to make items moddable anyway, so maybe we should just differentiate between BaseItem and ProceduralItem and then provide `src/callbacks/item/` with subdirs `onHit` `onInteract` `onUse` `printTooltip` 
The issue is that if we only have callbacks instead of making this fully code we cannot use the Data pattern, it worked for modifiers only because modifiers have typical Vtable layout. So you will have to stash that Data definition somewhere anyway and it cannot be in callback files because you would have to write it 3 times.
Now the trick is that modifiers also kind of want to have access to onHit and alike, so you have to propagate the calls there for stuff like #2834

**Comment by Wunka** [comment]
Maybe I can generate thorugh comptime magic a union out of all different ProceduralItem Type stats. Then when a modifier wants to access for example `strength` it can go through `@hasField` 
probagating should not be too difficult, probably a system like the windows use with onClick. so that an Item can say if it should probagate or not

**Comment by Argmaster** [comment]
Well, if you make an Union you should be able to just access it directly, but there will be no syntax highlighting which sucks.
But if we go for a union somewhere we should just use the structures contained in it directly.
(Sorry for closing, GH UI on phone is moody sometimes)

**Comment by Wunka** [comment]
> Well, if you make an Union you should be able to just access it directly, but there will be no syntax highlighting which sucks. But if we go for a union somewhere we should just use the structures contained in it directly.

The only reason I meant using `@hasFiield` is for example if a modifier can be used not only on for example pickaxes but also rings, but rings don't have `strength `so instead of crashing or creating a switch inside the modifier (which stops mods again) it can use `@hasField` to look if `strength `is there

**Comment by Argmaster** [comment]
Yeah that's a valid use case somewhat, tho I feel like our ideas on how to implement it are slightly different. It's fine tho.

**Comment by Wunka** [comment]
> Yeah that's a valid use case somewhat, tho I feel like our ideas on how to implement it are slightly different. It's fine tho.

I am probably also talking a lot of bulshit currently as I am not that experienced with such systems. I probably just need to try to actually write code for that and then the actual structure will be created through reviews etc. But I do agree that it sucks to not have syntax highlighting

**Comment by careeoki** [comment]
Unrelated but we should do bracelets instead of rings so they can be displayed on the player better (a ring would be like 1 pixel)

**Comment by Wunka** [comment]
> Yeah that's a valid use case somewhat, tho I feel like our ideas on how to implement it are slightly different. It's fine tho.

Can you maybe try to explain to me once again what your idea would be? I am currently thinking about it and just want the second opinion before I start with a system that will not be good.

**Comment by Wunka** [comment]
Summary of the talk on Discord:

- rename `Tool` to `ProcduralItem` to mirror the `BaseItem`
- all "propierties" like `damage` and `maxDurability` get saved in an array of `f32` 
- to index these use the `ToolProperty` where every entry value has the index of where the property is saved inside the array
- adding more properties would then involve adding something to the enum
- items get callbacks for things like `onLeftClick` which to use can be specified in the zon of either the item or the procedualItem zon (for example for the selection_wand: `.onLeftClick = .selectPos` and for all tools `.onLeftClick = .breakBlock`)

**Comment by H41ogen** [comment]
there should be a onRightClick callback too

**Comment by Wunka** [comment]
> there should be a onRightClick callback too

That was also talked about: for now onLeftClick is needed to "refactor" the current system. There will probably be a lot of callbacks, as this will also be also used for accessories etc. So things like `onEquip`, `onTick`... (I am just talking out of my mind these are not fixed) will be added later but for now just first creating the system only `onLeftClick` will exist

**Comment by H41ogen** [comment]
onRightClick would be necessary for the selection wand

**Comment by IntegratedQuantum** [comment]
the selection wand is not a procedural item.
